spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/ChangelogIterator.java

```diff
@@ -0,0 +1,207 @@
+
+  private final Iterator<Row> rowIterator;
+  private final int changeTypeIndex;
+  private final List<Integer> identifierFieldIdx;
+
+  private Row cachedRow = null;
+  private int[] indicesForIdentifySameRow = null;
+
+  private ChangelogIterator(
+      Iterator<Row> rowIterator, int changeTypeIndex, List<Integer> identifierFieldIdx) {
+    this.rowIterator = rowIterator;
+    this.changeTypeIndex = changeTypeIndex;
+    this.identifierFieldIdx = identifierFieldIdx;
+  }
+
+  /**
+   * Creates an iterator for records of a changelog table.
+   *
+   * @param rowIterator the iterator of rows from a changelog table
+   * @param changeTypeIndex the index of the change type column
+   * @param identifierFieldIdx the indices of the identifier columns, which determine if rows are
+   *     the same
+   * @return a new {@link ChangelogIterator} instance concatenated with the null-removal iterator
+   */
+  public static Iterator<Row> iterator(
+      Iterator<Row> rowIterator, int changeTypeIndex, List<Integer> identifierFieldIdx) {
+    ChangelogIterator changelogIterator =
+        new ChangelogIterator(rowIterator, changeTypeIndex, identifierFieldIdx);
+    return Iterators.filter(changelogIterator, Objects::nonNull);
+  }
+
+  @Override
+  public boolean hasNext() {
+    if (cachedRow != null) {
+      return true;
+    }
+    return rowIterator.hasNext();
+  }
+
+  @Override
+  public Row next() {
+    // if there is an updated cached row, return it directly
+    if (cachedUpdateRecord(cachedRow)) {
+      Row row = cachedRow;
+      cachedRow = null;
+      return row;
+    }
+
+    Row currentRow = currentRow();
+
+    if (indicesForIdentifySameRow == null) {
+      indicesForIdentifySameRow = generateIndicesForIdentifySameRow(currentRow);
+    }
+
+    if (currentRow.getString(changeTypeIndex).equals(DELETE) && rowIterator.hasNext()) {
+      GenericRowWithSchema nextRow = (GenericRowWithSchema) rowIterator.next();
+      cachedRow = nextRow;
+
+      if (isUpdateOrCarryoverRecord(currentRow, nextRow)) {
+        if (isCarryoverRecord(currentRow, nextRow)) {
+          // set carry-over rows to null for filtering out later
+          currentRow = null;
+          cachedRow = null;
+        } else {
+          Row[] rows = createUpdateChangelog((GenericRowWithSchema) currentRow, nextRow);
+          currentRow = rows[0];
+          cachedRow = rows[1];
+        }
+      }
+    }
+
+    return currentRow;
+  }
+
+  private int[] generateIndicesForIdentifySameRow(Row row) {
+    int[] indices = new int[row.length() - 1];
+    for (int i = 0; i < indices.length; i++) {
+      if (i < changeTypeIndex) {
+        indices[i] = i;
+      } else {
+        indices[i] = i + 1;
+      }
+    }
+    return indices;
+  }
+
+  private Row[] createUpdateChangelog(
+      GenericRowWithSchema currentRow, GenericRowWithSchema nextRow) {
+    GenericInternalRow deletedRow = new GenericInternalRow(currentRow.values());
```

Comment:
We should avoid introducing `InternalRows` just to call `update` on it because the internal representation is completely different (e.g. `UTF8String` vs `String`).

Also, we shouldn't limit ourselves to `GenericRowWithSchema`. We can have a more optimal implementation if we encounter it but we can't assume we will always get `GenericRowWithSchema`.